### ISSUE TITLE
MCMC in loop gives duplicate values

### ISSUE THREAD

**Issue by Binbose** [tensorflow/probability#1651]
I am trying to sample from a custom distribution using tfp's No-U-Turn sampler (in jax). I want to show a progress bar, so I tried to draw the samples in a loop, each time initializing the chain with the samples from the last iteration. But I seem to do it wrong, as I get many duplicate samples. (Also, a side question, what are the differences between the parallel_iteraions in the NoUTurnSampler and the sample_chain function, and how does it differ from initializing the state with a batch dimension > 1?)

Here is my code so far :

```
@partial(jit, static_argnums=(2,3,4,5))
def run_chain(key, initial_state, logpdf, ndim, dt, parallel_iterations):

    kernel = NoUTurnSampler(logpdf, dt, parallel_iterations=parallel_iterations)
    samples, log_probs = sample_chain(num_results=1,
                                  current_state=initial_state,
                                  kernel=kernel,
                                  trace_fn=lambda _, results: results.target_log_prob,
                                  num_burnin_steps=burning_samples,
                                  num_steps_between_results=0,
                                  parallel_iterations=parallel_iterations,
                                  seed=key)

    return samples.reshape(-1, ndim)


# Initial_samples
samples = (random.uniform(key, (n_samples, ndim,)) - 1 / 2)
all_samples = []
for _ in tqdm.trange(int(np.ceil(n_samples / parallel_chains))):
    samples = run_chain(split_key, samples, logpdf_vec, ndim, dt, parallel_chains)
    rng_key, split_key = jax.random.split(rng_key)
    all_samples.append(samples)
    
samples = np.concatenate(all_samples) # Has some duplicate samples
```

**Comment by ColCarroll** [comment]
NUTS has a [Metropolis acceptance step](https://en.wikipedia.org/wiki/Metropolis%E2%80%93Hastings_algorithm#Intuition) -- are you certain you are not just seeing a number of rejections?

**Comment by csuter** [comment]
Right, try making your step size (`dt`) smaller and see if you get more accepts

**Comment by csuter** [comment]
(you'll probably want to use a step size adaptation kernel at some point, to automate this during a warmup/burnin phase -- you might eventually want to look into using the `windowed_adaptive_nuts` sampler in `tfp.experimental.mcmc`)

**Comment by Binbose** [comment]
You are right; that solved it; thanks! 
I know this is off-topic from the initial question, but can you let me know what the difference between the `parallel_iterations` arguments in `NoUTurnSampler` and `sample_chain` and the `batch_size > 1` in the initial state is? The documentation seems to indicate that all of these causes parallel runs of several chains, but I don't understand what the behaviour is, is the number of parallel chains multiplied, or is the lowest/highest value taken or something else?

**Comment by csuter** [comment]
`parallel_iterations` is a pass-through to [`tf.while_loop`](https://www.tensorflow.org/api_docs/python/tf/while_loop) in the TF backend. It [appears](https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/internal/backend/numpy/control_flow.py#L77) to be unused in the jax substrate. In TF I think it's mostly useful for PRNG determinism, when using stateful samplers (moot in JAX).

**Comment by ColCarroll** [comment]
Closing for tidiness, but feel free to continue the conversation if you'd like!

**Comment by csuter** [comment]
Strictly speaking, number of chains equals the number of distinct values (batch shape) returned from your `target_log_prob_fn`. This will generally be the same as the batch shape of `current_state`, but it depends on how exactly the target log_prob fn is written, etc.

**Comment by csuter** [comment]
(to be clear, i can't think of any case where you'd want those shapes to be different, but maybe i'm overlooking something)
